### ISSUE TITLE
Kong Error: attempt to index field 'api' (a nil value)

### ISSUE THREAD

**Issue by jeremyjpj0916** [Kong/kong#3256]
EDIT - See my last comment for new steps to reproduce. 

### Summary

When Kong starts up for me every time now, the statsD message.api is not being populated, causing npe's with plugins such as the statsD or http logger. Anything in core been changed recently this week?

Well.. it cannot be that, I run 3 different environments DEV/STAGE/PROD. I only see this in one of our environments. Any ideas what could cause this or if you can reproduce it? 

### Steps To Reproduce

1. Use statsD plugin on your API, you should start to see NPE's right away.

### Additional Details & Logs

```
2018/02/26 05:33:58 [debug] 40#0: *413 [lua] base_plugin.lua:36: log(): executing plugin "statsd": log
2018/02/26 05:33:58 [notice] 40#0: *413 [lua] handler.lua:17: log(): {"request":{"querystring":{},"size":"76","uri":"\/F5\/status","url":"https:\/\/gateway-core.optum.com:8443\/F5\/status","headers":{"host":"gateway-core.optum.com","connection":"Close"},"method":"GET"},"client_ip":"10.1.51.1","latencies":{"request":0,"kong":0,"proxy":-1},"upstream_uri":"","response":{"headers":{"connection":"close","content-type":"application\/json; charset=utf-8","transfer-encoding":"chunked"},"status":404,"size":"213"},"started_at":1519623238832} while logging request, client: 10.1.51.1, server: kong, request: "GET /F5/status HTTP/1.1", host: "gateway-core.optum.com"
2018/02/26 05:33:58 [error] 40#0: *450 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/plugins/statsd/handler.lua:88: attempt to index field 'api' (a nil value)
stack traceback:
coroutine 0:
	/usr/local/share/lua/5.1/kong/plugins/statsd/handler.lua: in function </usr/local/share/lua/5.1/kong/plugins/statsd/handler.lua:83>, context: ngx.timer, client: 10.1.51.1, server: 0.0.0.0:8443
2018/02/26 05:33:58 [info] 44#0: *445 client closed connection while waiting for request, client: 10.1.20.1, server: 0.0.0.0:8443
```

Seems to be breaking here in handler(https://github.com/Kong/kong/blob/master/kong/plugins/statsd/handler.lua):

```
 local api_name = string_gsub(message.api.name, "%.", "_")
```

- Kong version 0.12.1
- Operating System: Docker Alpine build
- DB: Cassandra Cluster Multi DC 3 nodes in each with ONE for Read/Writes and Replication factor of 2.

**Comment by jeremyjpj0916** [comment]
Actually solved this, but I believe it to still be a sort of bug, in the event you call a proxy address that does not exist(I was missing a proxy in an environment I thought was there) you will see this error in your logs. I think it makes sense that if a proxy URI invoked does not exist then the request should be cut short before any plugins are called since many objects within them will have nil pointer exceptions thrown. Front end users still see: "{"message":"no API found with those values"}" but no reason to even invoke plugins on invalid proxy paths and eat up any processing time in plugins.

**Comment by thibaultcha** [comment]
This should have actually been fixed by f50d6546d757314d0ce5c24a8a9e1ac2b05737d2! THank you for the report nonetheless.

**Comment by jeremyjpj0916** [comment]
Ah I see:

```
+  else
+    -- TODO: this follows the pattern used by
+    -- https://github.com/Kong/kong/pull/2702 (which prevents an error from
+    -- being thrown and avoids confusing reports as per our metrics keys), but
+    -- as it stands, hides traffic from monitoring tools when the plugin is
+    -- configured globally. In fact, this basically disables this plugin when
+    -- it is configured to run globally, or per-consumer without an
+    -- API/Route/Service.
+    ngx_log(ngx.DEBUG,
+            "[statsd] no Route/Service/API in context, skipping logging")
+    return
```

I see this was applied to datadog as well. I wonder if the tcp/udp/http loggers are vulnerable to this as well when enabled globally. Glad to see it fixed for my statsD though! Onward to more bug squashing! Great work guys.
